### PR TITLE
Publish artifacts to Maven Central

### PR DESCRIPTION
**Pull request by wborn**
* Add artifact publishing and signing to Gradle build
* Add publishing steps to GHA workflow

With these changes GHA will also check if PRs build successfully.

Snapshot artifacts are published to Maven Central (https://s01.oss.sonatype.org/content/repositories/snapshots/io/openremote/) on pushes or workflow dispatches on the main branch.

When a release is created the workflow will publish the artifacts to a staging repository on Maven Central (https://s01.oss.sonatype.org/#stagingRepositories).

### CODE REVIEW

**Comment by richturner** [comment]
@wborn I had to delete the secrets I added specifically the `SIGNING_KEY` as it broke the github action (specifically the code that unifies the secrets and inputs it looks like), see here:

https://github.com/openremote/openremote/actions/runs/8662288322/job/23753942442?pr=1289

File: ORLib/build.gradle [hunk 1]
```diff
@@ -2,8 +2,12 @@ plugins {
     id 'com.android.library'
     id 'kotlin-android'
     id 'org.jetbrains.kotlin.android'
+    id 'maven-publish'
+    id 'signing'
 }
 
+version = hasProperty("orlibVersion") ? orlibVersion : "1.2.0-SNAPSHOT"
+
```
Comment (richturner) [inline ORLib/build.gradle:10]: Same comment here about version

**Review by richturner — COMMENTED** [review]
See code comments

File: GenericApp/app/build.gradle [hunk 2]
```diff
@@ -1,6 +1,12 @@
...
-apply plugin: 'com.google.gms.google-services'
+plugins {
+    id 'com.android.application'
+    id 'kotlin-android'
+    id 'com.google.gms.google-services'
+    id 'maven-publish'
+    id 'signing'
+}
+
+version = hasProperty("appVersion") ? appVersion : "1.2.0-SNAPSHOT"
```
[the remark this answers is not in the document]
Comment (wborn) [reply GenericApp/app/build.gradle:9]: The issue is that snapshot builds need a proper version. That way you can keep track of which snapshot version of the ORLib you have been testing with. Especially when producing snapshot versions on multiple branches e.g. 1.2.1-SNAPSHOT for a hotfix and 1.3.0-SNAPSHOT for the next bigger release.

Maybe it is best to add these versions to `gradle.properties` instead and update them before/after release? Such updates can then be done by the GHA release workflow.

That's similar to how it's normally done on Maven projects. The best practice is that release builds are reproducible. So in that case you don't want to provide version numbers as environment variables but have them as part of the files in the repo.

[the remark this answers is not in the document]
Comment (richturner) [reply GenericApp/app/build.gradle:9]: I can understand having them in the codebase for reproducability but the downside is remembering to change them but an automated process will mitigate this risk.

Obviously we should minimise in how many places such values need to be set and indeed inheriting from `gradle.properties` makes sense, I believe this is already the case for the main openremote repo.

[the remark this answers is not in the document]
Comment (wborn) [reply GenericApp/app/build.gradle]: I've updated the PR so it now uses [axion-release-plugin](https://github.com/allegro/axion-release-plugin/) for determining the version numbers.

File: ORLib/build.gradle (hunk 1 above)
Comment (wborn) [reply ORLib/build.gradle]: I've updated the PR so it now uses [axion-release-plugin](https://github.com/allegro/axion-release-plugin/) for determining the version numbers.

File: .github/workflows/ci_cd.yml [hunk 3]
```diff
@@ -32,15 +44,46 @@ jobs:
...
+        run: |
+          gradle assemble
+
+      - name: Run assemble and publish
+        if: ${{ steps.is_main_repo.outputs.value && github.event_name != 'pull_request' }}
         run: |
-          cd GenericApp
-          gradle build
+          echo "$KEYSTORE_BASE64" | base64 -d > keystore
+          command="gradle assemble publishToSonatype closeSonatypeStagingRepository"
```
Comment (wborn) [inline .github/workflows/ci_cd.yml:65]: When we're confident everything works well, we can close and release the artifacts immediately using:

```suggestion
          command="gradle assemble publishToSonatype closeAndReleaseSonatypeStagingRepository"
```

**Pull request merged**